CrimeMonitor/CrimeMonitor.cs

```diff
@@ -0,0 +1,849 @@
+                    List<FactionReport> reports = record.factionReports
+                        .Where(r => r.bounty && r.crimeDef != Crime.None && r.crimeDef != Crime.Bounty && r.shipId == @event.shipid)
+                        .ToList();
+
+                    if (reports?.Any() ?? false)
+                    {
+                        long total = reports.Sum(r => r.amount);
+
+                        // Check for discrepancy in logged bounties incurred
+                        if (total < @event.amount)
+                        {
+                            // Adjust the discrepancy report & remove when zeroed out
+                            FactionReport report = record.factionReports
+                                .FirstOrDefault(r => r.crimeDef == Crime.Bounty);
+                            if (report != null)
+                            {
+                                report.amount -= Math.Min(@event.amount - total, report.amount);
+                                if (report.amount == 0) { reports.Add(report); }
+                            }
+                        }
+                        // Remove associated bounties
+                        record.factionReports = record.factionReports.Except(reports).ToList();
+                    }
+                    // Adjust the total bounties incurred amount
+                    record.bounties -= Math.Min(@event.amount, record.bounties);
+
+                    RemoveRecord(record);
+                    update = true;
+                }
+            }
+            return update;
+        }
+
+        private void handleFineIncurredEvent(FineIncurredEvent @event)
+        {
+            if (@event.timestamp > updateDat)
+            {
+                updateDat = @event.timestamp;
+                _handleFineIncurredEvent(@event);
+                writeRecord();
+            }
+        }
+
+        private void _handleFineIncurredEvent(FineIncurredEvent @event)
+        {
+            int shipId = EDDI.Instance?.CurrentShip?.LocalId ?? 0;
+            Crime crime = Crime.FromEDName(@event.crimetype);
+            string currentSystem = EDDI.Instance?.CurrentStarSystem?.name;
+            FactionReport report = new FactionReport(@event.timestamp, false, shipId, crime, currentSystem, @event.fine)
+            {
+                station = EDDI.Instance?.CurrentStation?.name,
+                body = EDDI.Instance?.CurrentStellarBody?.name,
+                victim = @event.victim
+            };
+
+            FactionRecord record = GetRecordWithFaction(@event.faction);
+            if (record == null)
+            {
+                record = AddRecord(@event.faction);
+            }
+            record.factionReports.Add(report);
+            record.fines += report.amount;
+        }
+
+        private void handleFinePaidEvent(FinePaidEvent @event)
+        {
+            if (@event.timestamp > updateDat)
```

Comment:
See above. Move this if statement to the PreHandle method (rather than repeating it for each individual event)?

Reply:
Concur. Done.